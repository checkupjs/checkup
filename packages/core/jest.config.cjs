module.exports = {
  testEnvironment: 'node',
  moduleFileExtensions: ['ts', 'js', 'json'],
  transform: { '\\.ts$': 'ts-jest' },
  coverageReporters: ['lcov', 'text-summary'],
  // collectCoverage: !!process.env.CI,
  collectCoverageFrom: ['src/**/*.ts'],
  coveragePathIgnorePatterns: ['/templates/'],
  coverageThreshold: {
    global: {
      branches: 100,
      functions: 100,
      lines: 100,
      statements: 100,
    },
  },
  setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
  snapshotFormat: {
    printBasicPrototype: false,
  },
  testPathIgnorePatterns: ['/__fixtures__/', '/__utils__/'],
  extensionsToTreatAsEsm: ['.ts'],
  globals: {
    'ts-jest': {
      useESM: true,
    },
  },
  moduleNameMapper: {
    '^(\\.{1,2}/.*)\\.js$': '$1',
  },
};
