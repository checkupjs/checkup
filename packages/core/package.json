{
  "name": "@checkup/core",
  "description": "Checkup's core library",
  "version": "0.0.0",
  "author": "Steve Calvert <steve.calvert@gmail.com>",
  "bugs": "https://github.com/checkupjs/checkup/issues",
  "dependencies": {},
  "devDependencies": {
    "@types/chai": "^4",
    "@types/jest": "^25.1.1",
    "@types/node": "^13",
    "chai": "^4.2.0",
    "jest": "^25.1.0",
    "stdout-stderr": "^0.1.13",
    "ts-jest": "^25.2.0",
    "ts-node": "^8",
    "type-fest": "^0.10.0",
    "typescript": "^3.3"
  },
  "engines": {
    "node": "10.* || >= 12.*"
  },
  "files": [
    "/lib"
  ],
  "homepage": "https://github.com/checkupjs/checkup",
  "license": "MIT",
  "main": "lib/index.js",
  "repository": "https://github.com/checkupjs/checkup",
  "scripts": {
    "build": "tsc",
    "build:watch": "yarn build -w",
    "test": "jest"
  },
  "types": "lib/index.d.ts"
}
