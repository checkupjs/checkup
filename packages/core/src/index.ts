export { default as BaseTask } from './base-task';
export { default as BaseTaskResult } from './base-task-result';
export { default as FileSearcherTask } from './file-searcher-task';
export { default as FileSearcher } from './searchers/file-searcher';
export { default as CardData } from './pdf-components/card-data';
export { loadPlugins } from './loaders/plugin-loader';
export { getPackageJson } from './utils/get-package-json';
export { default as CheckupConfigService } from './configuration/checkup-config-service';
export { default as getSearchLoader } from './configuration/loaders/get-search-loader';
export { default as getFilepathLoader } from './configuration/loaders/get-filepath-loader';
export { ui } from './utils/ui';
export { toPairs } from './utils/data-transformers';

export * from './types';
