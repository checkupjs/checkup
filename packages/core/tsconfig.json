{
  "compilerOptions": {
    "declaration": true,
    "composite": true,
    "importHelpers": true,
    "module": "commonjs",
    "outDir": "lib",
    "rootDir": "src",
    "strict": true,
    "target": "es2017",
    "types": ["chai", "jest", "node"]
  },
  "include": ["src/**/*"]
}
