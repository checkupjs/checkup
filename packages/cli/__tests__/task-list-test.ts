import { Category, Priority, Task, TaskResult } from '@checkup/core';

import MockTask from './__utils__/mock-task';
import MockTaskResult from './__utils__/mock-task-result';
import TaskList from '../src/task-list';

//#region

class InsightsTaskHigh implements Task {
  meta = {
    taskName: 'insights-task-high',
    friendlyTaskName: 'Insights Task High',
    taskClassification: {
      category: Category.Insights,
      priority: Priority.High,
    },
  };

  async run(): Promise<TaskResult> {
    return new MockTaskResult(this.meta, 'insights task high is being run');
  }
}

class InsightsTaskLow implements Task {
  meta = {
    taskName: 'insights-task-low',
    friendlyTaskName: 'Insights Task Low',
    taskClassification: {
      category: Category.Insights,
      priority: Priority.Low,
    },
  };

  async run(): Promise<TaskResult> {
    return new MockTaskResult(this.meta, 'insights task low is being run');
  }
}

class RecommendationsTaskHigh implements Task {
  meta = {
    taskName: 'recommendations-task-high',
    friendlyTaskName: 'Recommendations Task High',
    taskClassification: {
      category: Category.Recommendations,
      priority: Priority.High,
    },
  };

  async run(): Promise<TaskResult> {
    return new MockTaskResult(this.meta, 'recommendations task high is being run');
  }
}

class RecommendationsTaskLow implements Task {
  meta = {
    taskName: 'recommendations-task-low',
    friendlyTaskName: 'Recommendations Task Low',
    taskClassification: {
      category: Category.Recommendations,
      priority: Priority.Low,
    },
  };

  async run(): Promise<TaskResult> {
    return new MockTaskResult(this.meta, 'recommendations task low is being run');
  }
}

class MigrationTaskHigh implements Task {
  meta = {
    taskName: 'migration-task-high',
    friendlyTaskName: 'Migration Task High',
    taskClassification: {
      category: Category.Migrations,
      priority: Priority.High,
    },
  };

  async run(): Promise<TaskResult> {
    return new MockTaskResult(this.meta, 'migration task high is being run');
  }
}

class MigrationTaskLow implements Task {
  meta = {
    taskName: 'migration-task-low',
    friendlyTaskName: 'Migration Task Low',
    taskClassification: {
      category: Category.Migrations,
      priority: Priority.Low,
    },
  };

  async run(): Promise<TaskResult> {
    return new MockTaskResult(this.meta, 'migration task low is being run');
  }
}
//#endregion

describe('TaskList', () => {
  it('can create an instance of a TaskList', () => {
    let taskList = new TaskList();

    expect(taskList).toBeInstanceOf(TaskList);
    expect(taskList.categories.size).toEqual(0);
  });

  it('registerTask adds a task to the TaskList', () => {
    let taskList = new TaskList();

    taskList.registerTask(new MockTask());

    expect(taskList.categories.get(Category.Insights)!.size).toEqual(1);
  });

  it('hasTask returns false if no task exists with that name', () => {
    let taskList = new TaskList();

    taskList.registerTask(new MockTask());

    expect(taskList.hasTask('foo')).toEqual(false);
  });

  it('hasTask returns true if task exists with that name', () => {
    let taskList = new TaskList();

    taskList.registerTask(new MockTask());

    expect(taskList.hasTask('mock-task')).toEqual(true);
  });

  it('findTask returns undefined if no task exists with that name', () => {
    let taskList = new TaskList();

    taskList.registerTask(new MockTask());

    expect(taskList.findTask('foo')).toBeUndefined();
  });

  it('findTask returns task instance if task exists with that name', () => {
    let taskList = new TaskList();

    taskList.registerTask(new MockTask());

    expect(taskList.findTask('mock-task')).toBeDefined();
  });

  it('runTask will run a task by taskName', async () => {
    let taskList = new TaskList();

    taskList.registerTask(new MockTask());

    let result = await taskList.runTask('mock-task');

    expect(result.json()).toMatchSnapshot();
  });

  it('runTasks will run all registered tasks', async () => {
    let taskList = new TaskList();

    taskList.registerTask(new MockTask());
    taskList.registerTask(new InsightsTaskHigh());

    let result = await taskList.runTasks();

    expect(result[0].json()).toMatchSnapshot();
    expect(result[1].json()).toMatchSnapshot();
  });

  it('runTasks will sort tasks in the correct order', async () => {
    let taskList = new TaskList();

    taskList.registerTask(new InsightsTaskLow());
    taskList.registerTask(new RecommendationsTaskHigh());
    taskList.registerTask(new MigrationTaskLow());
    taskList.registerTask(new MigrationTaskHigh());
    taskList.registerTask(new RecommendationsTaskLow());
    taskList.registerTask(new InsightsTaskHigh());

    let result = await taskList.runTasks();

    expect(result).toMatchInlineSnapshot(`
      Array [
        MockTaskResult {
          "meta": Object {
            "friendlyTaskName": "Insights Task High",
            "taskClassification": Object {
              "category": "insights",
              "priority": "high",
            },
            "taskName": "insights-task-high",
          },
          "result": "insights task high is being run",
        },
        MockTaskResult {
          "meta": Object {
            "friendlyTaskName": "Insights Task Low",
            "taskClassification": Object {
              "category": "insights",
              "priority": "low",
            },
            "taskName": "insights-task-low",
          },
          "result": "insights task low is being run",
        },
        MockTaskResult {
          "meta": Object {
            "friendlyTaskName": "Migration Task High",
            "taskClassification": Object {
              "category": "migrations",
              "priority": "high",
            },
            "taskName": "migration-task-high",
          },
          "result": "migration task high is being run",
        },
        MockTaskResult {
          "meta": Object {
            "friendlyTaskName": "Migration Task Low",
            "taskClassification": Object {
              "category": "migrations",
              "priority": "low",
            },
            "taskName": "migration-task-low",
          },
          "result": "migration task low is being run",
        },
        MockTaskResult {
          "meta": Object {
            "friendlyTaskName": "Recommendations Task High",
            "taskClassification": Object {
              "category": "recommendations",
              "priority": "high",
            },
            "taskName": "recommendations-task-high",
          },
          "result": "recommendations task high is being run",
        },
        MockTaskResult {
          "meta": Object {
            "friendlyTaskName": "Recommendations Task Low",
            "taskClassification": Object {
              "category": "recommendations",
              "priority": "low",
            },
            "taskName": "recommendations-task-low",
          },
          "result": "recommendations task low is being run",
        },
      ]
    `);
  });
});
