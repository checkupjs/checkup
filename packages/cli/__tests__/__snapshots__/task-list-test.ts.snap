// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList runTask will run a task by taskName 1`] = `
Object {
  "meta": Object {
    "friendlyTaskName": "Insights Task High",
    "taskClassification": Object {
      "category": "insights",
      "priority": "high",
    },
    "taskName": "insights-task-high",
  },
  "result": "insights task high is being run",
}
`;

exports[`TaskList runTasks will run all registered tasks 1`] = `
Object {
  "meta": Object {
    "friendlyTaskName": "Insights Task High",
    "taskClassification": Object {
      "category": "insights",
      "priority": "high",
    },
    "taskName": "insights-task-high",
  },
  "result": "insights task high is being run",
}
`;

exports[`TaskList runTasks will run all registered tasks 2`] = `
Object {
  "meta": Object {
    "friendlyTaskName": "Insights Task Low",
    "taskClassification": Object {
      "category": "insights",
      "priority": "low",
    },
    "taskName": "insights-task-low",
  },
  "result": "insights task low is being run",
}
`;
